source/validate.cpp

```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
   return SPV_SUCCESS;
 }
 
+// TODO(umar): Validate header
+spv_result_t setHeader(void* user_data, spv_endianness_t endian, uint32_t magic,
+                       uint32_t version, uint32_t generator, uint32_t id_bound,
+                       uint32_t reserved) {
+  (void)user_data;
+  (void)endian;
+  (void)version;
+  (void)generator;
+  (void)id_bound;
+  (void)reserved;
+  return SPV_SUCCESS;
+}
+
+// TODO(umar): Move this class to another file
+class validation_state_t {
+  const uint32_t* binary;
+  spv_diagnostic* diagnostic;
+  const uint32_t* program_counter;
+
+  std::set<uint32_t> declared_ids;
```

Comment:
Prefer std::unordered_set.  Regular sets are usually implemented as trees of nodes which use up a lot of space and are slow since there's so much pointer chasing.
(Just use the default hash function.)